importers/deb/deb.go

```diff
@@ -0,0 +1,301 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package deb implements deb package importer.
+package deb
+
+import (
+	"archive/tar"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+
+	"github.com/google/hashr/core/hashr"
+	"github.com/google/hashr/importers/common"
+
+	"pault.ag/go/debian/deb"
+)
+
+const (
+	// RepoName contains the repository name.
+	RepoName  = "deb"
+	chunkSize = 1024 * 1024 * 10 // 10MB
+)
+
+// Archive holds data related to deb archive.
+type Archive struct {
+	filename        string
+	remotePath      string
+	localPath       string
+	quickSha256hash string
+	repoPath        string
+}
+
+func subElem(parent, sub string) (bool, error) {
+	up := ".." + string(os.PathSeparator)
+
+	// path-comparisons using filepath.Abs don't work reliably according to docs (no unique representation).
+	rel, err := filepath.Rel(parent, sub)
+	if err != nil {
+		return false, err
+	}
+	if !strings.HasPrefix(rel, up) && rel != ".." {
+		return true, nil
+	}
+	return false, nil
+}
+
+func extractTar(tarfile *tar.Reader, outputFolder string) error {
+	for {
+		header, err := tarfile.Next()
+
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return fmt.Errorf("error while unpacking deb package: %v", err)
+		}
+
+		name := header.Name
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			continue
+		case tar.TypeReg:
+			unpackPath := filepath.Join(outputFolder, name)
+			unpackFolder := filepath.Dir(unpackPath)
+			fmt.Printf("a: %s, b: %s", unpackFolder, unpackPath)
+			if _, err := os.Stat(unpackFolder); os.IsNotExist(err) {
+				if err2 := os.MkdirAll(unpackFolder, 0755); err2 != nil {
+					return fmt.Errorf("error while creating target directory: %v", err2)
+				}
+			}
+
+			is_subelem, err := subElem(outputFolder, unpackPath)
```

Comment:
Can you please rename that is isSubelem?

Reply:
Done